code/cdw.py

```diff
@@ -0,0 +1,259 @@
+        else:
+            raise Exception('Unexpected nans! Aborting.')
+        
+        prep[:,:,:,inds] = this_prep
+    
+    bold_prep = nib.Nifti1Image(prep, bold.affine, bold.header)
+        
+    print('Done.')
+    return bold_prep
+
+def drop_by_labels(bold, labels, bad_labels):
+    """
+    Drop unwanted labels and fMRI frames
+    
+    Takes a 4D BOLD image and drops frames that correspond to unwanted
+    labels, and also drops those labels from the labels DataFrame.
+
+    Parameters
+    ----------
+    bold : Nifti1Image
+        Full 4D BOLD image.
+    labels : DataFrame
+        Full list of frame labels.
+    bad_labels : list of str
+        Unwanted labels.
+
+    Returns
+    -------
+    bold_subset : Nifti1Image
+        Selected BOLD frames.
+    labels_subset : DataFrame
+        Selected frame labels.
+
+    """
+    
+    good_inds = labels.query('labels not in @bad_labels').index
+    labels_subset = labels.loc[good_inds]
+    
+    print('Dropping unwanted frames...')
+    frames_subset = bold.get_fdata()[:,:,:,good_inds]
+    bold_subset = nib.Nifti1Image(frames_subset, bold.affine, bold.header)
+    
+    print('Done.')
+    return (bold_subset, labels_subset)
+    
+
+def reorder_by_labels(bold, labels):
+    """
+    Reorder labels and frames of 4D BOLD image
+    
+    Sorts labels (of frames) alphabetically and then reorders BOLD frames
+    to match that order.
+
+    Parameters
+    ----------
+    bold : Nifti1Image
+        Original 4D BOLD image.
+    labels : DataFrame
+        Original frame labels.
+
+    Returns
+    -------
+    bold_sorted : Nifti1Image
+        Sorted 4D BOLD image..
+    labels_sorted : DataFrame
+        Sorted frame labels.
+
+    """
+    
+    labels['fmri_i'] = range(len(labels))
+    labels_sorted = labels.sort_values(by=['labels','chunks'])
+    labels_sorted.reset_index(inplace=True)
+    
+    print('Reordering data...')
+    frames_sorted = bold.get_fdata()[:,:,:,labels_sorted['fmri_i']]
+    bold_sorted = nib.Nifti1Image(frames_sorted, bold.affine, bold.header)
+    
+    print('Done.')
+    return (bold_sorted, labels_sorted)
+
+
+def create_model_rdm(labels):
+    """
+    Create RDM from labels
+    
+    Creates a theoretical representational dissimilarity matrix from given
+    labels: dissimilarity within label is 0, between labels 1.
+
+    Parameters
+    ----------
+    labels : DataFrame
+        Labels of frames (sorted).
+
+    Returns
+    -------
+    rdm_model : ndarray
+        Theoretical RDM.
+
+    """
+    
+    # To get an "identity matrix" over labels we first expand labels to dummies
+    # (one column per label with values=[0,1]) and then get the dot product of
+    # that and its transpose -- result is 1 when the labels of the frame match,
+    # and 0 when they don't. To get RDM, switch those.
+    dummies = pd.get_dummies(labels.labels).to_numpy()
+    rdm_model = np.dot(dummies, dummies.T)
+    rdm_model = abs(rdm_model-1)
```

Comment:
This is not a direct translation of your intention at all. Your intention is to create an RDM that encodes the criterion that equal labels = 0 and unequal labels = 1. Although nice and efficient, this code is not an obvious implementation of that. We can do better. For example:

```python
def dissimilarity(label1, label2):
    return label1 != label2

rdm_model = squareform(pdist(labels.labels, metric=dissimilarity))
```